### ISSUE TITLE
SPA SSR considerations

### ISSUE THREAD

**Issue by xvandish** [livegrep/livegrep#326]
Hi @nelhage ! 

I was running into some performance issues with large `max_matches` values (like in #94) so I figured I'd try writing the UI in a more modern framework, to see if it could solve the perf pain points. 

I've started work on the UI as a Svelte rendered app ([SvelteKit](https://kit.svelte.dev/)), and good news - it solves the perf issues and is a super nice experience! Unfortunately, moving to any sort of SPA framework means we no longer get SSR as we do now, which seems a huge shame to lose. Maybe I'm getting too tied up in losing SSR, but if feels really nice to use websites when pages load pretty much instantly, without huge shifts in content, and it's so nice compared to the state of a lot of the web.

What do you think, is transitioning to the SPA without SSR worth the tradeoff? 
Pros:
1. Performance wins in render times pretty much everywhere
2. Much smaller bundle sizes - no need for backbone, jquery, bootstrap, etc
4. Dev experience for the UI itself is much nicer (hot-module reloading with Vite). 
5. Easy to serve, same as now the `/` handler is served with the FileHandler (or Nginx does it)

Cons:
1. No SSR!! Users will see blank pages until the JS loads then some API call is made made (this is mostly on the fileviewer / directory explorer). Issue can be mitigated by pre-loading pages, and it works pretty well, but this only helps after the initial load.

I thought a bit about what the architecture would look like if we did try to get SSR, and from what I can see, any sort of non-brittle SSR JS experience involves a Node server somewhere, and I can't think of a way to do it without introducing complexity or completely changing things. Here's a diagram of what I'm envisioning
```
[SPA] --- /api or /someAsset -> [Node server] ----- proxy /api, api calls for SSR --> [Go server]
```
Random thoughts: 
* `cmd/livegrep.go` should probably spin up both the node and go servers. 
* The Node server should be able to function without a reverse proxy in front of it, so it should itself proxy `/api` requests to the Go backend. 
* Any functionality within the Go api server that's proxy or header related needs to be replicated/shifted to the Node server...
* We now have to add Node to the docker images 
* It's just one more process/thing to have to consider -_-
* On the completely changing things, rewriting the entire server to be in Node is possible, but just need to make sure the Node grpc functionality is seamless. Can Bazel even compile proto to Nodejs atm?

I can also just try to solve the performance issues within the current infrastructure... I've also added in a bunch of little tweaks (`/` jumps to search bar, I fixed the left-jump of content when you start typing into the search bar, fixed up some text wrapping issues, etc) but there should be no problem in implementing those in the current infrastructure as well.

**Comment by nelhage** [comment]
I'm pretty enthusiastic about rewriting the UI into something more modern; The current thing is over a decade old and pretty janky. Especially if you have a prototype I'm happy pursuing that direction.

Regressing the initial page-load experience does seem slightly sad, but I'd be hopeful we can live with it without, at least to start. How bad is the initial render/reflow jank/delay in your current one?

Also, I haven't used Svelte much in anger, but (a) I know it's actually a compiler, not just a client-side framework and (b) our homepage is pretty static; I wonder if there's a model where we spin up a server during build and snapshot just the `/` path in such a way that we can load the initial HTML in a way that Svelte can take it and update from there. Pushing complexity to the build over the serving process seems much more preferable to me if something like that is at all an option.

**Comment by xvandish** [comment]
Cool, sounds good! I do have a prototype, it's actually pretty feature-complete - here's a video of what it feels/looks like when querying the torvalds/linux index - side by side with current experience. *typography/styles still super up in the air, let me know if there's anything you see that you're against/for.

https://user-images.githubusercontent.com/23644345/156861226-d4ed155e-873b-42fd-85b0-f36b54254fd2.mov


https://user-images.githubusercontent.com/23644345/156861663-49bbd37c-4a3d-48b5-94b6-5f582fb35db2.mov
*  prism styling, sometimes works, sometimes it doesn't. Haven't really looked into it yet.

The biggest change to the UI is the style/utility for deduping results for the same repo-path. Here's the same file shown side by side. The server  now (at a `/v2` route atm) does the deduping, so the client is only responsible for rendering.
<img width="1902" alt="Screen Shot 2022-03-04 at 6 03 41 PM" src="https://user-images.githubusercontent.com/23644345/156861805-872f824c-5829-44e6-b970-e393d57f6c32.png">


So, good news on the SSR front, I've been experimenting a bit, and even without SSR, I think we can avoid flashes of partial content! - SvelteKit has a [`load`](https://kit.svelte.dev/docs/loading) function that pages can use - "This function runs both during server-side rendering and in the client, and allows you to fetch and manipulate data before the page is rendered, thus preventing loading spinners." I can't show that for pure client side yet, since the dev server lets the page ssr.

So for the main `/` page, we get the search bar and all those components loaded in at once, because I make a `load` call that calls a `getServerInfo` endpoint (returns like internalUrls, available backends etc). This gives us a psuedo SSR like experience, which I think is definitely good enough for now! I've done the same thing for the file pages. So to answer your question, it actually feels exactly like the experience now (caveat that this is running of the dev server, haven't tested what it feels like when the Go server is serving the built artifacts).

Hmm, agreed that pushing to the build process would be preferable - I'll look into hydration options, but don't know off the top of my head at the moment.

Also, a bit of neutral/bad news - for huge `max_results` (like 2000/3000) etc Svelte barely has an edge, it's basically imperceptible. I'll see if I can optimize somewhere. I'll add in more timing info as I experiment.

Thanks for your time & sorry if I rambled!

**Comment by xvandish** [comment]
Since the huge `max_results` performance was disappointing in the Svelte version, I started experimenting with a hybrid approach where the server renders the search  results (using another go template) and the svelte client just renders the html string  like `{@html serverSearchResponse}` - and the early results are promising - it feels much faster (I'm working on a good way to get accurate performance timing) than both the svelte rewrite and current backbone version doing the rendering themselves. I think since we trust the api returning the markup (which is just us) we should be fine in regards to xss.. Any concerns at the outset in terms of going in this direction? Also, since the giant html payload sizes will get much smaller once compressed, this should be even faster in prod

Some data from querying the linux/torvalds repo. DOM update time was grabbed using Svelte's [`afterUpdate`](https://svelte.dev/docs#run-time-svelte-afterupdate)
| Framework | Query      | Network Time | Payload Size (not gziped) | Server Time | DOM Update Time | Total Time (network + DOM update)|
| -----------| ----------- | ----------- | ----------- | ----------- | ----------- | ----------|
|Svelte Hybrid| testing      | 31ms       | 170kb | 14ms | 7.599ms | 38.99ms
|Svelte Hybrid| testing max_matches:100      | 50ms       | 350kb | 26ms | 10.699ms | 60.699ms
|Svelte Hybrid| testing max_matches:200      | 58ms       | 664kb | 40ms | 22.099ms| 80.099ms
|Svelte Hybrid| testing max_matches:300      | 87ms       | 983kb | 65ms | 27.099ms| 114.099ms
|Svelte Hybrid| testing max_matches:1000      | 215ms       | 3.1MB | 183ms | 86.199ms | 301.199ms
|Svelte Hybrid| testing max_matches:2000      | 442ms       | 6.2MB | 409ms | 242.4ms | 684ms
|Svelte Hybrid| testing max_matches:3000      | 698ms       | 9.2MB | 663ms | 372.1ms | 1070.1ms
|Svelte Hybrid| testing max_matches:10000      | 1.14s       | 10.1MB | 1.1ms | 528.1 | 1688.1ms

Once I get a good way of doing performance comparisons I'll update with findings.

**Comment by xvandish** [comment]
@nelhage I've made the SSR changes above in my fork in [this pr](https://github.com/xvandish/livegrep/pull/32). Screenshots of new UI are in the PR.

I think some good things came out of it: namely
1. Can drop all of tools_dropbox_node, since I swapped webpack with esbuild (via the go api and Bazel genrules). Builds of the web portion are now ~12ms instead of ~5s.
2. Displaying searchResults is a matter of doing `searchResultsContainer.innerHtml = htmlResponse`, so all logic resides on the server
3. The client side no longer needs to keep a list of internal/external urls, making keeping up with a deployment where the index/repos available changes easier (don't have to keep web js inline with server)
4. Makes the initial search page payload much smaller (especially with a huge number of repos) 
5. Stripped out jquery,backbone and html.js - We now (after gzipping) only have 4kb of JS in total! Thats down from ~48kb from before.
6. Includes a new interface that inlines the search options in the search bar.

I'll close this issue out, if you're interested in any of this just let me know and I can piecemeal it to the upstream.
